intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+print(imgs.refine_names(..., 'H', 'W').names)
+
+def catch_error(fn):
+    try:
+        fn()
+    except RuntimeError as err:
+        print(err)
+
+# Tried to refine an existing name to a different name
+print(catch_error(lambda: imgs.refine_names('batch', 'channel', 'height', 'width')))
+
+######################################################################
+# Most simple operations propagate names. The ultimate goal for named tensors is
+# for all operations to propagate names in a reasonable, intuitive manner. Many
+# common operations have been added at the time of the 1.3 release; here,
+# for example, is `.abs()`:
+
+named_imgs = imgs.refine_names('N', 'C', 'H', 'W')
+print(named_imgs.abs().names)
+
+######################################################################
+# Accessors and Reduction
+# -----------------------
+#
+# One can use dimension names to refer to dimensions instead of the positional
+# dimension. These operations also propagate names. Indexing (basic and
+# advanced) has not been implemented yet but is on the roadmap. Using the `named_imgs`
+# tensor from above, we can do:
+
+output = named_imgs.sum(['C'])  # Perform a sum over the channel dimension
+print(output.names)
+
+img0 = named_imgs.select('N', 0)  # get one image
+print(img0.names)
+
+######################################################################
+# Name inference
+# --------------
+#
+# Names are propagated on operations in a two step process called **name inference**. It
+# works as follows:
+#
+# 1. **Check names**: an operator may check that certain dimensions must match.
+# 2. **Propagate names**: name inference computes and propagates output names to
+#    output tensors.
+#
+# Let's go through the very small example of adding 2 one-dim tensors with no
+# broadcasting.
+
+x = torch.randn(3, names=('X',))
+y = torch.randn(3)
+z = torch.randn(3, names=('Z',))
+
+# **Check names**: first, we will check whether the names of these two tensors
+# match. Two names match if and only if they are equal (string equality) or at
+# least one is ``None`` (``None``s are essentially a special wildcard name).
+# The only one of these three that will error, therefore, is ``x+z``:
+
+catch_error(lambda: x + z)
+
+# **Propagate names**: _unify_ the two names by returning the most refined name of
+# the two. With ``x + y``,  ``X`` is more specific than ``None``.
+
+print((x + y).names)
+
+######################################################################
+# Most name inference rules are straightforward but some of them (the dot
+# product ones) can have unexpected semantics. Let's go through a few more of
+# them.
+#
+# Broadcasting
+# ------------
+#
+# Named tensors do not change broadcasting behavior; they still broadcast by
+# position. However, when checking two dimensions for if they can be
+# broadcasted, the names of those dimensions must match. Two names match if and
+# only if they are equal (string equality), or if one is None.

```

Comment:
Remove "Two names match if and only if they are equal (string equality), or if one is None."